tensorflow/core/kernels/non_max_suppression_op.cc

```diff
@@ -195,6 +223,216 @@ void DoNonMaxSuppressionOp(
+      for (int i = 0; i < class_scores_data.size(); ++i) {
+        if (class_scores_data[i] > score_threshold) {
+          candidate_vector.emplace_back(Candidate({i, class_scores_data[i]}));
+        }
+      }
+
+      std::vector<int> selected;
+      std::vector<float> selected_boxes;
+      Candidate next_candidate;
+
+      std::sort(candidate_vector.begin(), candidate_vector.end(), cmp);
+      const Tensor const_boxes = boxes;
+      typename TTypes<float, 2>::ConstTensor boxes_data =
+          const_boxes.tensor<float, 2>();
+      int candidate_idx = 0;
+      while (selected.size() < size_per_class &&
+             candidate_idx < candidate_vector.size()) {
+        next_candidate = candidate_vector[candidate_idx++];
+
+        // Overlapping boxes are likely to have similar scores,
+        // therefore we iterate through the previously selected boxes backwards
+        // in order to see if `next_candidate` should be suppressed.
+        bool should_select = true;
+        for (int j = selected.size() - 1; j >= 0; --j) {
+          if (IOUGreaterThanThreshold(boxes_data, next_candidate.box_index,
+                                      selected[j], iou_threshold)) {
+            should_select = false;
+            break;
+          }
+        }
+
+        if (should_select) {
+          selected.push_back(next_candidate.box_index);
+          // Add the selected box to the result candidate. Sorted by score
+          int id = next_candidate.box_index;
+          ResultCandidate rc = {next_candidate.box_index,
+                                next_candidate.score,
+                                class_idx,
+                                {boxes_data(id, 0), boxes_data(id, 1),
+                                 boxes_data(id, 2), boxes_data(id, 3)}};
+          result_candidate_vec.push_back(rc);
+        }
+      }
+    }
+
+    auto rc_cmp = [](const ResultCandidate rc_i, const ResultCandidate rc_j) {
+      return rc_i.score > rc_j.score;
+    };
+    std::sort(result_candidate_vec.begin(), result_candidate_vec.end(), rc_cmp);
+
+    int max_detections = 0;
+    // If pad_per_class is false, we always pad to max_total_size
+    if (!pad_per_class) {
+      max_detections =
+          std::min((int)result_candidate_vec.size(), total_size_per_batch);
+      per_batch_size = total_size_per_batch;
+    } else {
+      per_batch_size =
+          std::min(total_size_per_batch, max_size_per_class * num_classes);
+      max_detections =
+          std::min(per_batch_size, (int)result_candidate_vec.size());
+    }
+
+    final_valid_detections.push_back(max_detections);
+
+    int curr_total_size = max_detections;
+    int result_idx = 0;
+    // Pick the top max_detections values
+    while (curr_total_size > 0 && result_idx < result_candidate_vec.size()) {
+      ResultCandidate next_candidate = result_candidate_vec[result_idx++];
+      // Add to final output vectors
+      nmsed_boxes[batch].push_back(
```

Comment:
Maybe reserve this if haven't before: 
1) nmsed_boxes[batch] to be of size (4 * (predicted #iterations of the loop )).
2) nmsed_scores[batch] to be of size (predicted #iterations of the loop ).
3) nmsed_classes[batch] to be of size (predicted #iterations of the loop ).

Where size of iteration will be the minimum between min(curr_total_size, result_candidate_vec.size())